File: app/lib/shared/models/task.dart
class Task {
  static const String groupLabelPrefix = '分组:';

  final int id;
  final String name;
  final String command;
  final String cronExpression;
  final List<String> cronExpressions;
  final String taskType;
  final String pythonVersion;
  final double status;
  final String labels;
  final List<String> displayLabels;
  final DateTime? lastRunAt;
  final DateTime? nextRunAt;
  final int? lastRunStatus;
  final int timeout;
  final int? randomDelaySeconds;
  final int maxRetries;
  final int retryInterval;
  final List<int> successExitCodes;
  final bool notifyOnFailure;
  final bool notifyOnSuccess;
  final bool notifyOnAbort;
  final int? notificationChannelId;
  final int? dependsOn;
  final int sortOrder;
  final bool isPinned;
  final String? taskBefore;
  final String? taskAfter;
  final bool allowMultipleInstances;
  final String? notificationChannelName;
  final bool? notificationChannelEnabled;
  final double? lastRunningTime;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Task({
    required this.id,
    required this.name,
    required this.command,
    required this.cronExpression,
    this.cronExpressions = const [],
    this.taskType = 'cron',
    this.pythonVersion = '3.12',
    required this.status,
    this.labels = '',
    this.displayLabels = const [],
    this.lastRunAt,
    this.nextRunAt,
    this.lastRunStatus,
    this.timeout = 0,
    this.randomDelaySeconds,
    this.maxRetries = 0,
    this.retryInterval = 0,
    this.successExitCodes = const [0],
    this.notifyOnFailure = false,
    this.notifyOnSuccess = false,
    this.notifyOnAbort = false,
    this.notificationChannelId,
    this.dependsOn,
    this.sortOrder = 0,
    this.isPinned = false,
    this.taskBefore,
    this.taskAfter,
    this.allowMultipleInstances = false,
    this.notificationChannelName,
    this.notificationChannelEnabled,
    this.lastRunningTime,
    required this.createdAt,
    required this.updatedAt,
  });

  bool get isDisabled => status == 0;
  bool get isQueued => status == 0.5;
  bool get isEnabled => status == 1;
  bool get isRunning => status == 2;

  String get statusText {
    if (isRunning) return '运行中';
    if (isQueued) return '排队中';
    if (isEnabled) return '已启用';
    if (isDisabled) return '已禁用';
    return '未知状态（${_numberText(status)}）';
  }

  String get taskTypeText {
    switch (taskType) {
      case 'cron':
        return '常规定时';
      case 'manual':
        return '手动运行';
      case 'startup':
        return '开机运行';
      default:
        return '未知类型（$taskType）';
    }
  }

  String get lastRunStatusText {
    switch (lastRunStatus) {
      case null:
        return '未运行';
      case 0:
        return '成功';
      case 1:
        return '失败';
      case 2:
        return '已终止';
      default:
        return '未知结果（$lastRunStatus）';
    }
  }

  List<String> get labelList => labels.isEmpty
      ? []
      : labels.split(',').where((l) => l.isNotEmpty).toList();

  List<String> get labelsForDisplay =>
      displayLabels.isNotEmpty ? displayLabels : labelList;

  List<String> get effectiveCronExpressions => cronExpressions.isNotEmpty
      ? List.unmodifiable(cronExpressions)
      : cronExpression.trim().isEmpty
      ? const []
      : [cronExpression.trim()];

  static List<String> parseCronInput(String value) => value
      .split(RegExp(r'\r?\n'))
      .map((expression) => expression.trim())
      .where((expression) => expression.isNotEmpty)
      .toList();

  static Map<String, dynamic> cronPayload(String value) {
    final expressions = parseCronInput(value);
    return {
      'cron_expression': expressions.isEmpty ? '' : expressions.first,
      'cron_expressions': expressions,
    };
  }

  static bool isGroupLabel(String label) =>
      label.trim().startsWith(groupLabelPrefix);

  static String toGroupLabel(String group) =>
      '$groupLabelPrefix${group.trim()}';

  static List<int>? parseSuccessExitCodes(String value) {
    final parts = value
        .split(RegExp(r'[,，\s]+'))
        .where((part) => part.isNotEmpty);
    final result = <int>[];
    final seen = <int>{};
    for (final part in parts) {
      final code = int.tryParse(part);
      if (code == null || code < 0 || code > 255) return null;
      if (seen.add(code)) result.add(code);
    }
    return result.isEmpty ? null : result;
  }

  String? get groupName {
    for (final label in labelList) {
      final trimmed = label.trim();
      if (isGroupLabel(trimmed)) {
        final group = trimmed.substring(groupLabelPrefix.length).trim();
        if (group.isNotEmpty) {
          return group;
        }
      }
    }
    return null;
  }

  List<String> get userLabelsForDisplay {
    final visible = labelsForDisplay
        .where((label) => !isGroupLabel(label))
        .toList();
    final group = groupName;
    if (group != null && group.isNotEmpty) {
      visible.remove(group);
    }
    return visible;
  }

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: _int(json['id']),
      name: json['name']?.toString() ?? '',
      command: json['command']?.toString() ?? '',
      cronExpression: json['cron_expression']?.toString() ?? '',
      cronExpressions: json['cron_expressions'] is List
          ? (json['cron_expressions'] as List)
                .map((e) => e.toString().trim())
                .where((s) => s.isNotEmpty)
                .toList()
          : const [],
      taskType: json['task_type']?.toString() ?? 'cron',
      pythonVersion: json['python_version']?.toString() ?? '3.12',
      status: _double(json['status']),
      labels: json['labels'] is List
          ? (json['labels'] as List).join(',')
          : json['labels']?.toString() ?? '',
      displayLabels: json['display_labels'] is List
          ? (json['display_labels'] as List)
                .map((e) => e.toString())
                .where((label) => label.trim().isNotEmpty)
                .toList()
          : const [],
      lastRunAt: _date(json['last_run_at']),
      nextRunAt: _date(json['next_run_at']),
      lastRunStatus: _intOrNull(json['last_run_status']),
      timeout: _int(json['timeout']),
      randomDelaySeconds: _intOrNull(json['random_delay_seconds']),
      maxRetries: _int(json['max_retries']),
      retryInterval: _int(json['retry_interval']),
      successExitCodes: _intList(json['success_exit_codes'], fallback: const [0]),
      notifyOnFailure: json['notify_on_failure'] == true,
      notifyOnSuccess: json['notify_on_success'] == true,
      notifyOnAbort: json['notify_on_abort'] == true,
      notificationChannelId: _intOrNull(json['notification_channel_id']),
      dependsOn: _intOrNull(json['depends_on']),
      sortOrder: _int(json['sort_order']),
      isPinned: json['is_pinned'] == true,
      taskBefore: json['task_before']?.toString(),
      taskAfter: json['task_after']?.toString(),
      allowMultipleInstances: json['allow_multiple_instances'] == true,
      notificationChannelName: json['notification_channel_name']?.toString(),
      notificationChannelEnabled: json['notification_channel_enabled'] as bool?,
      lastRunningTime: _doubleOrNull(json['last_running_time']),
      createdAt: _date(json['created_at']) ?? DateTime.now(),
      updatedAt: _date(json['updated_at']) ?? DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() => {
    'name': name,
    'command': command,
    'cron_expression': effectiveCronExpressions.isEmpty
        ? ''
        : effectiveCronExpressions.first,
    'cron_expressions': effectiveCronExpressions,
    'task_type': taskType,
    'python_version': pythonVersion,
    'labels': labels,
    'timeout': timeout,
    'random_delay_seconds': randomDelaySeconds,
    'max_retries': maxRetries,
    'retry_interval': retryInterval,
    'success_exit_codes': successExitCodes.join(','),
    'notify_on_failure': notifyOnFailure,
    'notify_on_success': notifyOnSuccess,
    'notify_on_abort': notifyOnAbort,
    'notification_channel_id': notificationChannelId,
    'depends_on': dependsOn,
    'sort_order': sortOrder,
    'task_before': taskBefore,
    'task_after': taskAfter,
    'allow_multiple_instances': allowMultipleInstances,
  };
}

int _int(dynamic v) => _intOrNull(v) ?? 0;
int? _intOrNull(dynamic v) => v is num
    ? v.toInt()
    : int.tryParse(v?.toString().trim() ?? '');
double _double(dynamic v) => _doubleOrNull(v) ?? 0.0;
double? _doubleOrNull(dynamic v) => v is num
    ? v.toDouble()
    : double.tryParse(v?.toString().trim() ?? '');
List<int> _intList(dynamic value, {List<int> fallback = const []}) {
  if (value is String) {
    return Task.parseSuccessExitCodes(value) ?? fallback;
  }
  if (value is! List) return fallback;
  final parsed = Task.parseSuccessExitCodes(value.join(','));
  return parsed ?? fallback;
}
String _numberText(num value) => value == value.toInt()
    ? value.toInt().toString()
    : value.toString();
DateTime? _date(dynamic v) {
  if (v is String && v.isNotEmpty) return DateTime.tryParse(v);
  return null;
}

File: panel/server/model/task.go
package model

import (
	"fmt"
	"strconv"
	"strings"
	"time"
	"unicode"
)

const (
	TaskStatusDisabled = 0
	TaskStatusQueued   = 0.5
	TaskStatusEnabled  = 1
	TaskStatusRunning  = 2

	TaskTypeCron    = "cron"
	TaskTypeManual  = "manual"
	TaskTypeStartup = "startup"

	RunSuccess = 0
	RunFailed  = 1
	RunAborted = 2

	DefaultSuccessExitCodes = "0"
)

type Task struct {
	ID             uint    `gorm:"primarykey" json:"id"`
	Name           string  `gorm:"size:128;not null" json:"name"`
	Command        string  `gorm:"type:text;not null" json:"command"`
	PythonVersion  string  `gorm:"size:16;default:''" json:"python_version"`
	CronExpression string  `gorm:"type:text;not null" json:"cron_expression"`
	TaskType       string  `gorm:"size:16;not null;default:'cron'" json:"task_type"`
	Status         float64 `gorm:"not null" json:"status"`
	// LastStartupAutoRunDate 只记录“开机运行”的自动触发日期，手动运行不受它限制。
	LastStartupAutoRunDate string     `gorm:"size:10;default:''" json:"last_startup_auto_run_date"`
	Labels                 string     `gorm:"size:256;default:''" json:"-"`
	LastRunAt              *time.Time `json:"last_run_at"`
	LastRunStatus          *int       `json:"last_run_status"`
	Timeout                int        `gorm:"default:0" json:"timeout"`
	SuccessExitCodes       string     `gorm:"size:128;not null;default:'0'" json:"success_exit_codes"`
	RandomDelaySeconds     *int       `json:"random_delay_seconds"`
	MaxRetries             int        `json:"max_retries"`
	RetryInterval          int        `json:"retry_interval"`
	NotifyOnFailure        bool       `json:"notify_on_failure"`
	NotifyOnSuccess        bool       `json:"notify_on_success"`
	NotifyOnAbort          bool       `gorm:"default:0" json:"notify_on_abort"`
	NotificationChannelID  *uint      `gorm:"index" json:"notification_channel_id"`
	DependsOn              *uint      `gorm:"index" json:"depends_on"`
	SortOrder              int        `json:"sort_order"`
	IsPinned               bool       `json:"is_pinned"`
	PID                    *int       `gorm:"column:pid" json:"pid"`
	LogPath                *string    `gorm:"size:256" json:"log_path"`
	LastRunningTime        *float64   `json:"last_running_time"`
	TaskBefore             *string    `gorm:"type:text" json:"task_before"`
	TaskAfter              *string    `gorm:"type:text" json:"task_after"`
	AllowMultipleInstances bool       `json:"allow_multiple_instances"`
	StopSchedule           string     `gorm:"type:text;default:''" json:"stop_schedule"`
	CreatedAt              time.Time  `json:"created_at"`
	UpdatedAt              time.Time  `json:"updated_at"`
}

func (Task) TableName() string {
	return "tasks"
}

func (t *Task) ToDict() map[string]interface{} {
	labels := []string{}
	if t.Labels != "" {
		labels = strings.Split(t.Labels, ",")
	}
	cronExpressions := splitTaskCronExpressions(t.CronExpression)

	return map[string]interface{}{
		"id":                       t.ID,
		"name":                     t.Name,
		"command":                  t.Command,
		"python_version":           t.PythonVersion,
		"cron_expression":          t.CronExpression,
		"cron_expressions":         cronExpressions,
		"task_type":                t.GetTaskType(),
		"status":                   t.Status,
		"labels":                   labels,
		"last_run_at":              t.LastRunAt,
		"last_run_status":          t.LastRunStatus,
		"timeout":                  t.Timeout,
		"success_exit_codes":       t.GetSuccessExitCodes(),
		"random_delay_seconds":     t.RandomDelaySeconds,
		"max_retries":              t.MaxRetries,
		"retry_interval":           t.RetryInterval,
		"notify_on_failure":        t.NotifyOnFailure,
		"notify_on_success":        t.NotifyOnSuccess,
		"notify_on_abort":          t.NotifyOnAbort,
		"notification_channel_id":  t.NotificationChannelID,
		"depends_on":               t.DependsOn,
		"sort_order":               t.SortOrder,
		"is_pinned":                t.IsPinned,
		"pid":                      t.PID,
		"log_path":                 t.LogPath,
		"last_running_time":        t.LastRunningTime,
		"task_before":              t.TaskBefore,
		"task_after":               t.TaskAfter,
		"allow_multiple_instances": t.AllowMultipleInstances,
		"stop_schedule":            t.StopSchedule,
		"created_at":               t.CreatedAt,
		"updated_at":               t.UpdatedAt,
	}
}

// NormalizeSuccessExitCodes 统一任务表单、导入文件和旧数据中的成功退出码格式。
// Shell 退出码只接受 0-255；负数保留给超时、信号退出等面板失败状态，不能配置为成功。
func NormalizeSuccessExitCodes(raw string) (string, error) {
	parts := strings.FieldsFunc(raw, func(r rune) bool {
		return r == ',' || r == '，' || unicode.IsSpace(r)
	})
	if len(parts) == 0 {
		return DefaultSuccessExitCodes, nil
	}

	seen := make(map[int]struct{}, len(parts))
	normalized := make([]string, 0, len(parts))
	for _, part := range parts {
		code, err := strconv.Atoi(part)
		if err != nil || code < 0 || code > 255 {
			return "", fmt.Errorf("成功退出码只能填写 0-255 的整数，多个值请用逗号分隔")
		}
		if _, exists := seen[code]; exists {
			continue
		}
		seen[code] = struct{}{}
		normalized = append(normalized, strconv.Itoa(code))
	}
	return strings.Join(normalized, ","), nil
}

func (t *Task) GetSuccessExitCodes() string {
	if t == nil {
		return DefaultSuccessExitCodes
	}
	normalized, err := NormalizeSuccessExitCodes(t.SuccessExitCodes)
	if err != nil {
		// 旧库或手工改库出现异常值时回退到标准退出码 0，不能扩大成功范围。
		return DefaultSuccessExitCodes
	}
	return normalized
}

func (t *Task) IsSuccessExitCode(exitCode int) bool {
	if t == nil || exitCode < 0 {
		return false
	}
	target := strconv.Itoa(exitCode)
	for _, code := range strings.Split(t.GetSuccessExitCodes(), ",") {
		if code == target {
			return true
		}
	}
	return false
}

func splitTaskCronExpressions(raw string) []string {
	lines := strings.FieldsFunc(raw, func(r rune) bool {
		return r == '\n' || r == '\r'
	})
	result := make([]string, 0, len(lines))
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line != "" {
			result = append(result, line)
		}
	}
	return result
}

func NormalizeTaskType(taskType string) string {
	switch strings.ToLower(strings.TrimSpace(taskType)) {
	case "", TaskTypeCron:
		return TaskTypeCron
	case TaskTypeManual:
		return TaskTypeManual
	case TaskTypeStartup:
		return TaskTypeStartup
	default:
		return ""
	}
}

func IsValidTaskType(taskType string) bool {
	return NormalizeTaskType(taskType) != ""
}

func (t *Task) GetTaskType() string {
	if t == nil {
		return TaskTypeCron
	}
	return NormalizeTaskType(t.TaskType)
}

func (t *Task) UsesCronSchedule() bool {
	return t.GetTaskType() == TaskTypeCron
}

func (t *Task) SetLabelsFromSlice(labels []string) {
	t.Labels = strings.Join(labels, ",")
}

func (t *Task) GetLabels() []string {
	if t.Labels == "" {
		return []string{}
	}
	return strings.Split(t.Labels, ",")
}
